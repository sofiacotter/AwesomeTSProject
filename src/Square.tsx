import React, {useEffect} from 'react';
import {Pressable, Text} from 'react-native';
import Animated, {
  useAnimatedStyle,
  useSharedValue,
  withRepeat,
  withSpring,
  withTiming,
} from 'react-native-reanimated';
import styles from './stylesheet';

/* Function components: Componentes que apenas contém método render()
e não mantêm estado próprio. Em vez de serem definidos como uma classe
que estende de React.Component, são funções que incluem props como input
e retornam o que deve ser renderizado */

/*É importante fazer assign de keys em listas, para definir a sua ordem
e as operações e remover, adicionar, substituir... as keys não têm de ser
únicas globalmnte, mas sim entre componentes e os seus "irmãos".
<li key={this.props.id}> */

/* Variáveis imutáveis: não substituir as variáveis diretamente, mas sim
por uma cópia das mesmas, alterada */

interface SquareProps {
  isWin: boolean | null;
  value: null | string;
  onClick: () => void;
}

const Square = ({isWin, value, onClick}: SquareProps): JSX.Element => {
  //const progress = useSharedValue(1);
  const scale = useSharedValue(2);

  const animatedStyles = useAnimatedStyle(() => {
    return {
      //opacity: progress.value,
      transform: [{scale: scale.value}],
    };
  }, []);

  const playAnimation = (): void => {
    scale.value = withRepeat(withSpring(1), 2, true);
  };
  const combPress = (): void => {
    playAnimation();
    onClick();
  };

  //useEffect(() => {}, []);

  if (isWin) {
    return (
      //botão verde
      <Pressable style={styles.winsquare} onPress={combPress}>
        <Animated.Text style={(styles.text, animatedStyles)}>
          {value}
        </Animated.Text>
      </Pressable>
    );
  }
  return (
    //botão vermelho
    <Pressable style={styles.square} onPress={combPress}>
      <Animated.Text style={(styles.text, animatedStyles)}>
        {value}
      </Animated.Text>
    </Pressable>
  );
};

export default Square;
