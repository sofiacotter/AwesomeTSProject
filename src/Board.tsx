import React from 'react';
import {View} from 'react-native';
import Square from './Square';
import stylesheet from './stylesheet';

interface BoardProps {
  squares: (null | string)[];
  lines: null | number[];
  onClick: (i: number) => void;
}

const Board = ({squares, lines, onClick}: BoardProps): JSX.Element => {
  const renderSquare = (i: number): JSX.Element => {
    const isWinningSquare: boolean | null = lines && lines.includes(i);
    return (
      <Square
        key={i}
        value={squares[i]}
        isWin={isWinningSquare}
        onClick={() => {
          onClick(i);
        }}
      />
    );
  };
  //Tentar fazer esta função usar os for loops
  const createSquares = () => {
    let rows: JSX.Element[] = [];
    for (var i = 0; i < 3; i++) {
      let squaresArray: JSX.Element[] = [];
      for (var j = 0; j < 3; j++) {
        squaresArray.push(renderSquare(3 * i + j));
      }
      rows.push(
        <View style={stylesheet.linha} key={i}>
          {squaresArray}
        </View>,
      );
    }
    return rows;
  };

  return <View style={stylesheet.coluna}>{createSquares()}</View>;
};

export default Board;
