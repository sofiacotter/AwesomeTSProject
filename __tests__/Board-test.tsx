import React from 'react';
import renderer from 'react-test-renderer';
import Square from '../src/Square';
import {fireEvent, render, screen, within} from '@testing-library/react-native';
import Board from '../src/Board';
import {View} from 'react-native';

describe('Testing Board Properties', () => {
  //TESTAR A RENDERIZAÇÃO DO BOARD
  /*
  test('Square Test', () => {
    const tree = renderer
      .create(
        <Square
          isWin={false}
          value={'X'}
          onClick={function (): void {
            throw new Error('Function not implemented.');
          }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  */

  //CARREGAR NUM SQUARE NULL E ELE MUDAR PARA 'X'
  test('Square should have particular values', () => {
    function testClick(i: number): void {
      console.log('CLICOU NO SQUARE: ', i);
    }
    const isWin = false;
    const iSquare = 1;
    const squares = Array(9).fill(null);
    const renderSquare = (i: number): JSX.Element => {
      return (
        <Square
          key={i}
          value={squares[i]}
          isWin={isWin}
          onClick={() => {
            testClick(iSquare);
          }}
        />
      );
    };
    const createSquares = () => {
      let rows: JSX.Element[] = [];
      for (var i = 0; i < 3; i++) {
        let squares_array: JSX.Element[] = [];
        for (var j = 0; j < 3; j++) {
          squares_array.push(renderSquare(3 * i + j));
        }
        rows.push(<View key={i}>{squares_array}</View>);
      }
      return rows;
    };
    render(<View>{createSquares()}</View>);

    const SquareButton = screen.queryAllByDisplayValue('o');
    fireEvent(SquareButton, 'click');
  });

  //CARREGAR NUM SQUARE 'X' E ELE NÃO FAZ NADA
  //CARREGAR NUM SQUARE EM LINHA DE 3 E ELES MUDAREM A COR PARA VERDE
});
