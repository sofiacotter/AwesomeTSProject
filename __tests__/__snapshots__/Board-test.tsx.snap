// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Board Properties Square Test 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#f5fffa",
      "borderColor": "#b0e0e6",
      "borderWidth": 3,
      "flex": 1,
      "fontWeight": "bold",
      "justifyContent": "center",
      "margin": 2,
      "padding": "5%",
      "textAlign": "center",
    }
  }
  testID="squarePressId"
>
  <Text
    style={
      Object {
        "transform": Array [
          Object {
            "scale": 2,
          },
        ],
      }
    }
    testID="textId"
  >
    X
  </Text>
</View>
`;
